sgl-kernel/csrc/moe/moe_fused_gate.cu

```diff
@@ -0,0 +1,493 @@
+#include <ATen/cuda/CUDAContext.h>
+#include <stdio.h>
+#include <torch/extension.h>
+
+#include <cfloat>
+#include <type_traits>
+
+#ifndef USE_ROCM
+#include <cuda_runtime.h>
+#include <cutlass/array.h>
+#include <cutlass/cutlass.h>
+#include <cutlass/numeric_types.h>
+template <typename T, int N>
+using AlignedArray = cutlass::AlignedArray<T, N>;
+using bfloat16_t = cutlass::bfloat16_t;
+using float16_t = cutlass::half_t;
+using float32_t = float;
+#else
+// Native aligned array template with operator[] support as a substitution of cutlass::AlignedArray
+template <typename T, int N>
+struct AlignedArray {
+  T data[N];
+
+  __device__ __host__ T& operator[](int i) {
+    return data[i];
+  }
+  __device__ __host__ const T& operator[](int i) const {
+    return data[i];
+  }
+};
+using bfloat16_t = at::BFloat16;
+using float16_t = at::Half;
+using float32_t = float;
+#endif
+
+// QQ NOTE: to handle the case for at::Half, error: more than one operator ">" matches these operands: built-in operator
+// "arithmetic > arithmetic" function "operator>(const __half &, const __half &)"
+template <typename T>
+__device__ inline bool cmp_gt(const T& a, const T& b) {
+  if constexpr (std::is_same<T, at::Half>::value) {
+    // at::Half (or float16_t in our native case) causes ambiguity, so we cast to float.
+    return static_cast<float>(a) > static_cast<float>(b);
+  } else {
+    // For types like float, at::BFloat16, or cutlass::half_t / cutlass::bfloat16_t, assume operator> works as expected.
+    return a > b;
+  }
+}
+
+template <typename T>
+__device__ inline bool cmp_eq(const T& a, const T& b) {
+  if constexpr (std::is_same<T, at::Half>::value) {
+    return static_cast<float>(a) == static_cast<float>(b);
+  } else {
+    return a == b;
+  }
+}
+
+#ifndef USE_ROCM
```

Comment:
@yiakwy-xpu-ml-framework-team I temporarily put this here to differentiate between cutlass one and native array with torch dtype. Feel free to let me know what you prefer to change since I might misunderstand some comments. Thank you!